File: tools/gluten-it/common/src/main/scala/org/apache/gluten/integration/tpc/action/Parameterized.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.gluten.integration.tpc.action

import org.apache.gluten.integration.stat.RamStat
import org.apache.gluten.integration.tpc.{TpcRunner, TpcSuite}
import org.apache.spark.sql.ConfUtils.ConfImplicits._
import org.apache.spark.sql.SparkSessionSwitcher
import org.apache.commons.lang3.exception.ExceptionUtils
import org.apache.gluten.integration.tpc.action.Actions.QuerySelector

import scala.collection.immutable.Map
import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer

class Parameterized(
    scale: Double,
    queries: QuerySelector,
    explain: Boolean,
    iterations: Int,
    warmupIterations: Int,
    configDimensions: Seq[Dim],
    excludedCombinations: Seq[Set[DimKv]],
    metrics: Array[String])
  extends Action {

  private def validateDims(configDimensions: Seq[Dim]): Unit = {
    if (
      configDimensions
        .map(
          dim => {
            dim.name
          })
        .toSet
        .size != configDimensions.size
    ) {
      throw new IllegalStateException("Duplicated dimension name found")
    }

    configDimensions.foreach {
      dim =>
        if (dim.dimValues.map(dimValue => dimValue.name).toSet.size != dim.dimValues.size) {
          throw new IllegalStateException("Duplicated dimension value found")
        }
    }
  }

  private val coordinates: mutable.LinkedHashMap[Coordinate, Seq[(String, String)]] = {
    validateDims(configDimensions)
    val dimCount = configDimensions.size
    val coordinateMap = mutable.LinkedHashMap[Coordinate, Seq[(String, String)]]()

    def fillCoordinates(
        dimOffset: Int,
        intermediateCoordinates: Map[String, String],
        intermediateConf: Seq[(String, String)]): Unit = {
      if (dimOffset == dimCount) {
        // we got one coordinate
        excludedCombinations.foreach {
          ec: Set[DimKv] =>
            if (ec.forall {
              kv =>
                intermediateCoordinates.contains(kv.k) && intermediateCoordinates(kv.k) == kv.v
            }) {
              println(s"Coordinate ${Coordinate(intermediateCoordinates)} excluded by $ec.")
              return
            }
        }
        coordinateMap(Coordinate(intermediateCoordinates)) = intermediateConf
        return
      }
      val dim = configDimensions(dimOffset)
      dim.dimValues.foreach {
        dimValue =>
          fillCoordinates(
            dimOffset + 1,
            intermediateCoordinates + (dim.name -> dimValue.name),
            intermediateConf ++ dimValue.conf)
      }
    }

    fillCoordinates(0, Map(), Seq())

    coordinateMap
  }

  override def execute(tpcSuite: TpcSuite): Boolean = {
    val runner: TpcRunner = new TpcRunner(tpcSuite.queryResource(), tpcSuite.dataWritePath(scale))
    val allQueries = tpcSuite.allQueryIds()

    val sessionSwitcher = tpcSuite.sessionSwitcher
    val testConf = tpcSuite.getTestConf()

    println("Prepared coordinates: ")
    coordinates.toList.map(_._1).zipWithIndex.foreach {
      case (c, idx) =>
        println(s"  $idx: $c")
    }
    coordinates.foreach {
      entry =>
        // register one session per coordinate
        val coordinate = entry._1
        val coordinateConf = entry._2
        val conf = testConf.clone()
        conf.setAllWarningOnOverriding(coordinateConf)
        sessionSwitcher.registerSession(coordinate.toString, conf)
    }

    val runQueryIds = queries.select(tpcSuite)

    // warm up
    (0 until warmupIterations).foreach {
      _ =>
        runQueryIds.foreach {
          queryId => Parameterized.warmUp(queryId, tpcSuite.desc(), sessionSwitcher, runner)
        }
    }

    val results = coordinates.flatMap {
      entry =>
        val coordinate = entry._1
        val coordinateResults = (0 until iterations).flatMap {
          iteration =>
            println(s"Running tests (iteration $iteration) with coordinate $coordinate...")
            runQueryIds.map {
              queryId =>
                Parameterized.runTpcQuery(
                  runner,
                  sessionSwitcher,
                  queryId,
                  coordinate,
                  tpcSuite.desc(),
                  explain,
                  metrics)
            }
        }.toList
        coordinateResults
    }

    val dimNames = configDimensions.map(dim => dim.name)

    val passedCount = results.count(l => l.succeed)
    val count = results.count(_ => true)

    // RAM stats
    println("Performing GC to collect RAM statistics... ")
    System.gc()
    System.gc()
    printf(
      "RAM statistics: JVM Heap size: %d KiB (total %d KiB), Process RSS: %d KiB\n",
      RamStat.getJvmHeapUsed(),
      RamStat.getJvmHeapTotal(),
      RamStat.getProcessRamUsed()
    )

    println("")
    println("Test report: ")
    println("")
    printf("Summary: %d out of %d queries passed. \n", passedCount, count)
    println("")
    TestResultLines(dimNames, metrics, results.filter(_.succeed)).print()
    println("")

    if (passedCount == count) {
      println("No failed queries. ")
      println("")
    } else {
      println("Failed queries: ")
      println("")
      TestResultLines(dimNames, metrics, results.filter(!_.succeed)).print()
      println("")
    }

    if (passedCount != count) {
      return false
    }
    true
  }
}

case class DimKv(k: String, v: String)
case class Dim(name: String, dimValues: Seq[DimValue])
case class DimValue(name: String, conf: Seq[(String, String)])
case class Coordinate(coordinate: Map[String, String]) // [dim, dim value]

case class TestResultLine(
    queryId: String,
    succeed: Boolean,
    coordinate: Coordinate,
    rowCount: Option[Long],
    executionTimeMillis: Option[Long],
    metrics: Map[String, Long],
    errorMessage: Option[String])

case class TestResultLines(
    dimNames: Seq[String],
    metricNames: Seq[String],
    lines: Iterable[TestResultLine]) {
  def print(): Unit = {
    var fmt = "|%15s|%15s"
    for (_ <- dimNames.indices) {
      fmt = fmt + "|%20s"
    }
    for (_ <- metricNames.indices) {
      fmt = fmt + "|%35s"
    }
    fmt = fmt + "|%30s|%30s|\n"
    val fields = ArrayBuffer[String]("Query ID", "Succeed")
    dimNames.foreach(dimName => fields.append(dimName))
    metricNames.foreach(metricName => fields.append(metricName))
    fields.append("Row Count")
    fields.append("Query Time (Millis)")
    printf(fmt, fields: _*)
    lines.foreach {
      line =>
        val values = ArrayBuffer[Any](line.queryId, line.succeed)
        dimNames.foreach {
          dimName =>
            val coordinate = line.coordinate.coordinate
            if (!coordinate.contains(dimName)) {
              throw new IllegalStateException("Dimension name not found" + dimName)
            }
            values.append(coordinate(dimName))
        }
        metricNames.foreach {
          metricName =>
            val metrics = line.metrics
            values.append(metrics.getOrElse(metricName, "N/A"))
        }
        values.append(line.rowCount.getOrElse("N/A"))
        values.append(line.executionTimeMillis.getOrElse("N/A"))
        printf(fmt, values: _*)
    }
  }
}

object Parameterized {
  private def runTpcQuery(
      runner: TpcRunner,
      sessionSwitcher: SparkSessionSwitcher,
      id: String,
      coordinate: Coordinate,
      desc: String,
      explain: Boolean,
      metrics: Array[String]) = {
    println(s"Running query: $id...")
    try {
      val testDesc = "Gluten Spark %s %s %s".format(desc, id, coordinate)
      sessionSwitcher.useSession(coordinate.toString, testDesc)
      runner.createTables(sessionSwitcher.spark())
      val result =
        runner.runTpcQuery(sessionSwitcher.spark(), testDesc, id, explain, metrics)
      val resultRows = result.rows
      println(
        s"Successfully ran query $id. " +
          s"Returned row count: ${resultRows.length}")
      TestResultLine(
        id,
        succeed = true,
        coordinate,
        Some(resultRows.length),
        Some(result.executionTimeMillis),
        result.metrics,
        None)
    } catch {
      case e: Exception =>
        val error = Some(s"FATAL: ${ExceptionUtils.getStackTrace(e)}")
        println(
          s"Error running query $id. " +
            s" Error: ${error.get}")
        TestResultLine(id, succeed = false, coordinate, None, None, Map.empty, error)
    }
  }

  private[tpc] def warmUp(
      id: String,
      desc: String,
      sessionSwitcher: SparkSessionSwitcher,
      runner: TpcRunner): Unit = {
    println(s"Warming up: Running query: $id...")
    try {
      val testDesc = "Gluten Spark %s %s warm up".format(desc, id)
      sessionSwitcher.useSession("test", testDesc)
      runner.createTables(sessionSwitcher.spark())
      val result = runner.runTpcQuery(sessionSwitcher.spark(), testDesc, id, explain = false)
      val resultRows = result.rows
      println(
        s"Warming up: Successfully ran query $id. " +
          s"Returned row count: ${resultRows.length}")
    } catch {
      case e: Exception =>
        val error = Some(s"FATAL: ${ExceptionUtils.getStackTrace(e)}")
        println(
          s"Warming up: Error running query $id. " +
            s" Error: ${error.get}")
    }
  }
}

File: tools/gluten-it/common/src/main/java/org/apache/gluten/integration/command/Parameterized.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.gluten.integration.command;

import org.apache.gluten.integration.BaseMixin;
import org.apache.gluten.integration.collections.JavaCollectionConverter;

import com.google.common.base.Preconditions;
import org.apache.commons.lang3.ArrayUtils;
import picocli.CommandLine;

import java.util.*;
import java.util.concurrent.Callable;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import scala.Tuple2;
import scala.collection.JavaConverters;

@CommandLine.Command(
    name = "parameterized",
    mixinStandardHelpOptions = true,
    showDefaultValues = true,
    description = "Run queries with parameterized configurations")
public class Parameterized implements Callable<Integer> {
  @CommandLine.Mixin private BaseMixin mixin;

  @CommandLine.Mixin private DataGenMixin dataGenMixin;

  @CommandLine.Mixin private QueriesMixin queriesMixin;

  @CommandLine.Option(
      names = {"--warmup-iterations"},
      description = "Dry-run iterations before actually run the test",
      defaultValue = "0")
  private int warmupIterations;

  @CommandLine.Option(
      names = {"-m", "--metric"},
      description = "Specify a series of executor metrics to collect during execution")
  private String[] metrics = new String[0];

  @CommandLine.Option(
      names = {"-d", "--dim"},
      description =
          "Set a series of dimensions consisting of possible config options, example: -d=offheap:1g,spark.memory.offHeap.enabled=true,spark.memory.offHeap.size=1g")
  private String[] dims = new String[0];

  @CommandLine.Option(
      names = {"--excluded-dims"},
      description =
          "Set a series of comma-separated list of dimension combinations to exclude. Example: --exclude-dims=offheap:1g,aqe=on")
  private String[] excludedDims = new String[0];

  private static final Pattern dimPattern1 =
      Pattern.compile("([\\w-]+):([^,:]+)((?:,[^=,]+=[^=,]+)*)");
  private static final Pattern dimPattern2 = Pattern.compile("([^,:]+)((?:,[^=,]+=[^=,]+)*)");

  private static final Pattern excludedDimsPattern =
      Pattern.compile("[\\w-]+:[^,:]+(?:,[\\w-]+:[^,:]+)*");

  @Override
  public Integer call() throws Exception {
    final Map<String, Map<String, List<Map.Entry<String, String>>>> parsed = new LinkedHashMap<>();

    final scala.collection.immutable.Seq<
            scala.collection.immutable.Set<
                org.apache.gluten.integration.action.Parameterized.DimKv>>
        excludedCombinations =
            JavaCollectionConverter.asScalaSeq(
                Arrays.stream(excludedDims)
                    .map(
                        d -> {
                          final Matcher m = excludedDimsPattern.matcher(d);
                          Preconditions.checkArgument(
                              m.matches(), "Unrecognizable excluded dims: " + d);
                          Set<org.apache.gluten.integration.action.Parameterized.DimKv> out =
                              new HashSet<>();
                          final String[] dims = d.split(",");
                          for (String dim : dims) {
                            final String[] kv = dim.split(":");
                            Preconditions.checkArgument(
                                kv.length == 2, "Unrecognizable excluded dims: " + d);
                            out.add(
                                new org.apache.gluten.integration.action.Parameterized.DimKv(
                                    kv[0], kv[1]));
                          }
                          return JavaConverters.asScalaSetConverter(out)
                              .asScala()
                              .<org.apache.gluten.integration.action.Parameterized.DimKv>toSet();
                        })
                    .collect(Collectors.toList()));

    // parse dims
    for (String dim : dims) {
      Matcher matcher1 = dimPattern1.matcher(dim);
      Matcher matcher2 = dimPattern2.matcher(dim);
      if (matcher1.matches() == matcher2.matches()) {
        throw new IllegalArgumentException("Unexpected dim: " + dim);
      }
      final String dimName;
      final String dimValueName;
      final String confText;

      if (matcher1.matches()) {
        dimName = matcher1.group(1);
        dimValueName = matcher1.group(2);
        confText = matcher1.group(3);
      } else {
        // matcher2.matches
        dimName = matcher2.group(1);
        dimValueName = matcher2.group(0);
        confText = matcher2.group(2);
      }

      final List<Map.Entry<String, String>> options = new ArrayList<>();
      final List<String> splits =
          Arrays.stream(confText.split(",")).filter(s -> !s.isEmpty()).collect(Collectors.toList());
      for (String split : splits) {
        String[] kv = split.split("=");
        if (kv.length != 2) {
          throw new IllegalArgumentException("Unexpected dim: " + dim);
        }
        options.add(new AbstractMap.SimpleImmutableEntry<>(kv[0], kv[1]));
      }

      parsed
          .computeIfAbsent(dimName, s -> new LinkedHashMap<>())
          .computeIfAbsent(dimValueName, s -> new ArrayList<>())
          .addAll(options);
    }

    // Convert Map<String, Map<String, List<Map.Entry<String, String>>>> to List<Dim>
    scala.collection.immutable.Seq<org.apache.gluten.integration.action.Parameterized.Dim>
        parsedDims =
            JavaCollectionConverter.asScalaSeq(
                parsed.entrySet().stream()
                    .map(
                        e ->
                            new org.apache.gluten.integration.action.Parameterized.Dim(
                                e.getKey(),
                                JavaCollectionConverter.asScalaSeq(
                                    e.getValue().entrySet().stream()
                                        .map(
                                            e2 ->
                                                new org.apache.gluten.integration.action
                                                    .Parameterized.DimValue(
                                                    e2.getKey(),
                                                    JavaCollectionConverter.asScalaSeq(
                                                        e2.getValue().stream()
                                                            .map(
                                                                e3 ->
                                                                    new Tuple2<>(
                                                                        e3.getKey(), e3.getValue()))
                                                            .collect(Collectors.toList()))))
                                        .collect(Collectors.toList()))))
                    .collect(Collectors.toList()));

    org.apache.gluten.integration.action.Parameterized parameterized =
        new org.apache.gluten.integration.action.Parameterized(
            queriesMixin.queries(),
            queriesMixin.explain(),
            queriesMixin.iterations(),
            warmupIterations,
            queriesMixin.noSessionReuse(),
            parsedDims,
            excludedCombinations,
            JavaCollectionConverter.asScalaSeq(Arrays.asList(metrics)));
    return mixin.runActions(ArrayUtils.addAll(dataGenMixin.makeActions(), parameterized));
  }
}
